### ISSUE TITLE
Rename To fails if new name only differs in char case

### ISSUE THREAD

**Issue by ProgerXP** [ProgerXP/Notepad2e#140]
Currently File > Rename To misbehaves when new file name is only different in character case. Issues:

1. Open File dialog displays overwrite prompt (e.g. old name a.txt, new name is A.txt). This is likely a Windows issue that we can't fix.
2. Submission results in the file being deleted altogether (due to MoveFile call failing to create a new file because of the same names and subsequent DeleteFile call deleting the old file). This needs to be fixed (only if new name is different in char case) by adding an intermediate step of renaming original file to a temporary file in the same folder + renaming the temp file into new file (with different char case). 

Please ensure the temporary file's name is unique (don't overwrite any existing file)! And don't create it elsewhere (e.g. in %temp%) to avoid leaking sensitive data to another medium.

**Comment by cshnik** [comment]
Done.

**Comment by ProgerXP** [comment]
```
GetTempFileName(szDirectory, L"tmp", rand() + 1, szMaxPathTmpFile)
```
Why are you using random number here rather than simply giving the function uUnique of `0` so that it automatically repeats attempts to create a non-existing file name **and** create that file (to avoid race condition after your function returns and before new file is created (renamed in our case)?

Using 0 will simplify the code by removing the checks, and you can use MoveFileEx with a flag to allow replacing existing file as we trust GetTempFileName won't return a name of an existing file that it hasn't itself created.

```
+        if (!n2e_RenameFileToTemporary(szCurFile, nMaxTryAttempts, szTmpFile))
+        {
+          DeleteFile(szCurFile);
```

Wait, does it read as "try renaming old file into a new file using an intermediate temp file and if that files - **simply delete the old file**" (!)?

**Comment by cshnik** [comment]
> Why are you using random number here rather than simply giving the function uUnique of 0 so that it automatically repeats attempts to create a non-existing file name and create that file (to avoid race condition after your function returns and before new file is created (renamed in our case)?

Fixed.

> Wait, does it read as "try renaming old file into a new file using an intermediate temp file and if that files - simply delete the old file" (!)?

Fixed.
Here are the original steps of "Rename To"-feature:
> 1. Select new file name
> 2. Save current file contents using new file name
> 3. Delete old file

Improved scheme of "Rename To"-feature's work:
> 1. Select new file name
> 2. If there was an overwrite prompt then try to generate temporary filename and rename current file to temporary file
> 3. Save current contents using new file name
> 4. Delete temporary file (if it was created on step 2)
> 5. If step 3 has failed (file was not saved) then temporary file renamed back to original

**Comment by ProgerXP** [comment]
I don't like the flow of FileSave at all, it's so tangled that it's hard to see what happens in all situations. The function should be rewritten with less branches and parameters so it's easy to understand:

* Some parameter combinations make no sense; such parameters must be merged into a single enum-type parameter. In particular, bDeleteOld assumes bSaveAs and !bSaveCopy because it requires a new file name from the user. Also, bSaveCopy assumes bSaveAs for the same reason.
* Instead of "delete old" flag use a more clear "rename/move old" flag. We never want to delete old file, we want it to be renamed - if we simply delete it this means losing user's data which is the last thing we want in a text editor (we already had issues with Open Next/Previous which discarded buffer changes; please don't introduce another such condition, this is a critical piece of code).
* I'd like to avoid any file dialog hooks. I understand how the hook is supposed to work but we already had numerous issues with Open By Prefix which bugs we fought for over a year; if the hook is necessary to workaround the first issue ("Open File dialog displays overwrite prompt...") then let's remove the hook and allow the dialog to produce overwrite prompt according to its default behaviour. I will tolerate this.
* To determine if a temporary file should be used, compare old and new file names after the dialog returns, instead of using the reserved field (which I also don't like) and having the comparison in the hook. Or, better, consider using a temporary file only after the regular move fails because I'm not sure we can trust any strcmp variant to ignore characters that are guaranteed to fail the move (does it fail if only ASCII chars are case-insensitively the same? or it's using locale settings?).
* Note that MoveFile is a poor choice here (probably used by the original author) because it doesn't work cross-volume. Better use MoveFileEx.
* Check result of important functions (such as MoveFile) and show a message if saving fails. It seems that currently such failures are silently ignored.

**Comment by cshnik** [comment]
> - I'd like to avoid any file dialog hooks. I understand how the hook is supposed to work but we already had numerous issues with Open By Prefix which bugs we fought for over a year; if the hook is necessary to workaround the first issue ("Open File dialog displays overwrite prompt...") then let's remove the hook and allow the dialog to produce overwrite prompt according to its default behaviour. I will tolerate this.
> - To determine if a temporary file should be used, compare old and new file names after the dialog returns, instead of using the reserved field (which I also don't like) and having the comparison in the hook. Or, better, consider using a temporary file only after the regular move fails because I'm not sure we can trust any strcmp variant to ignore characters that are guaranteed to fail the move (does it fail if only ASCII chars are case-insensitively the same? or it's using locale settings?).

Unfortunately, there is no way to extract/preserve selected case specific filename other than using of the implemented hook. The point is: the result returned by GetSaveFileName()-call after overwrite prompt accepted is always the same as originally used filename (current filename), e.g.:
1. Open file a.txt
2. Alt+F6, select A.TXT, submit SaveFile dialog
3. Press Yes on overwrite prompt
Actual result: GetSaveFileName() returns originally opened filename: a.txt as if it was submitted by user.

Implemented hook is actually used to preserve the selected filename (which does not returned/ignored after the overwrite prompt acceptance by default).

**Comment by ProgerXP** [comment]
I see. I have considered this issue. Let's revert the entire change. Renaming to a different char case is relatively rare operation but I don't like the implications this brings to the code.

After reverting I will review the FileSave function and we'll rewrite part of it because it was convoluted even in the original form. And we will add an extra check so that file is only deleted if move succeeded (possibly just use MoveFileEx for this).

**Comment by cshnik** [comment]
Recent changes were rolled back.

**Comment by ProgerXP** [comment]
Changes to FileSave in this fork are in fact smaller than I thought. The only added piece is this:

>        // [2e]: File->RenameTo menu item
>        if (bDeleteOld
>            && lstrlen(szCurFile)
>            && lstrcmp(szCurFile, tchFile)
>            )
>        {
>          DeleteFile(szCurFile);
>        }
>        // [/2e]

1. Find a way to determine if two strings (file paths) are referring to the same file. We could use `lstrcmpi` but I am worried it may not work with all locales - I want some function that's specifically working with file names and knows about Windows naming rules. Does WinAPI has an equivalent of Unix' inode/stat() that uniquely identifies a file record? NTFS has hardlinks so there must be a similar concept.
2. Insert a check after `SaveFileDlg` call that compares old file name with newly chosen file name using the above routine and, if they're identical, shows the following error message (and returns from `FileSave` as if the save dialog was cancelled):
```
New file name refers to the same old file. This is not allowed.
```
The following line is also appended, after an empty line, if `bDeleteOld` was set:
```
If you were trying to rename it into a different character case, this would not work due to Windows Save File Dialog limitation (#140). To work around this, do two renames using a temporary file name.
```
3. Remove `lstrcmp` check from the `if` as it's no more necessary (covered by item 2 above).

**Comment by cshnik** [comment]
> 1. Find a way to determine if two strings (file paths) are referring to the same file. We could use lstrcmpi but I am worried it may not work with all locales - I want some function that's specifically working with file names and knows about Windows naming rules.

I. There is no locale aspect for filenames since Notepad2e is built using unicode.
II. There is no need to check file uniqueness by filenames since user's provided filename never returned from the API-call (if there was an overwrite prompt):
>  The point is: the result returned by GetSaveFileName()-call after overwrite prompt accepted is always the same as originally used filename (current filename), e.g.:
> 1. Open file a.txt
> 2. Alt+F6, select A.TXT, submit SaveFile dialog
> 3. Press Yes on overwrite prompt
Actual result: GetSaveFileName() returns originally opened filename: a.txt as if it was submitted by user.

At this point ``lstrcmp`` is totally enough to detect if newly provided filename is correct:
1. <> 0 - provided filename is different from original one
1. 0 - provided filename is the same

Fixed. Please check current implementation and confirm if any additional changes are required.

**Comment by ProgerXP** [comment]
> The point is: the result returned by GetSaveFileName()-call after overwrite prompt accepted is always the same as originally used filename (current filename), e.g.:

This is not always the case:

1. Open the program, save document as `a.txt`
2. Save As `a.` (notice last period, this saves file without extension)
3. Open `a.txt`
4. Save As `a.`

You will see overwrite prompt in save file dialog but no warning message box.

This is what I mean by reliably determining if two path names refer to the same physical file node. We can't expect save dialog to work exactly as we need, sadly.

**Comment by cshnik** [comment]
>> The point is: the result returned by GetSaveFileName()-call after overwrite prompt accepted is always the same as originally used filename (current filename), e.g.:

> This is not always the case:
>
> 1. Open the program, save document as `a.txt`
> 2. Save As `a.` (notice last period, this saves file without extension)
> 3. Open `a.txt`
> 4. Save As `a.`
>
>You will see overwrite prompt in save file dialog but no warning message box.

The problem exists only when user provides the name which is different only in some chars case.
Original statement should be read as:
> The point is: **in case of this issue** the result returned by GetSaveFileName()-call after overwrite prompt accepted is always the same as originally used filename (current filename)

Regarding your example: there is no problem here, GetSaveFileName()-call returns new completely different filename and FileSave() will handle this in it's regular way.
Files ``a.txt`` and ``a.`` are different/independent even if they have the same content and are located in the same folder. It looks strange when working with ``a.txt`` user will get a warning box when trying to save file as ``a.``: ``New file name refers to the same old file. This is not allowed.``

> This is what I mean by reliably determining if two path names refer to the same physical file node. 

There is no scenario available within Notepad2e to make the same file(file instance) have different filenames/references.

**Comment by ProgerXP** [comment]
>>3. Open a.txt
>>4. Save As a.

Agree, this was my mistake. 

Let's hope the dialog really works as we expect and never returns a different string for the same file name.
